src/error.rs

```diff
@@ -0,0 +1,87 @@
+    CannotOpenFd,
+    /// File descriptor from the jobserver environment variable value is not a pipe.
+    NotAPipe,
+    /// Jobserver inheritance is not supported on this platform.
+    Unsupported,
+}
+
+impl FromEnvError {
+    /// Get public version of error type.
+    pub fn kind(&self) -> FromEnvErrorKind {
+        match self.inner {
+            FromEnvErrorInner::NoEnvVar => FromEnvErrorKind::NoEnvVar,
+            FromEnvErrorInner::CannotParse(_) => FromEnvErrorKind::CannotParse,
+            #[cfg(any(unix, windows))]
+            FromEnvErrorInner::CannotOpenPath(_, _) => FromEnvErrorKind::CannotOpenPath,
+            #[cfg(unix)]
+            FromEnvErrorInner::CannotOpenFd(_, _) => FromEnvErrorKind::CannotOpenFd,
+            #[cfg(unix)]
+            FromEnvErrorInner::NotAPipe(_, _) => FromEnvErrorKind::NotAPipe,
+            #[cfg(not(any(unix, windows)))]
+            FromEnvErrorInner::Unsupported => FromEnvErrorKind::Unsupported,
+        }
+    }
+}
+
+impl std::fmt::Display for FromEnvError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.inner {
+            FromEnvErrorInner::NoEnvVar => write!(f, "there is no environment variable that describes jobserver to inherit"),
+            FromEnvErrorInner::CannotParse(s) => write!(f, "cannot parse jobserver environment variable value: {s}"),
+            #[cfg(any(unix, windows))]
+            FromEnvErrorInner::CannotOpenPath(s, err) => write!(f, "cannot open path or name {s} from the jobserver environment variable value: {err}"),
+            #[cfg(unix)]
+            FromEnvErrorInner::CannotOpenFd(fd, err) => write!(f, "cannot open file descriptor {fd} from the jobserver environment variable value: {err}"),
+            #[cfg(unix)]
+            FromEnvErrorInner::NotAPipe(fd, None) => write!(f, "file descriptor {fd} from the jobserver environment variable value is not a pipe"),
+            #[cfg(unix)]
+            FromEnvErrorInner::NotAPipe(fd, Some(err)) => write!(f, "file descriptor {fd} from the jobserver environment variable value is not a pipe: {err}"),
+            #[cfg(not(any(unix, windows)))]
+            FromEnvErrorInner::Unsupported => write!(f, "jobserver inheritance is not supported on this platform"),
+        }
+    }
+}
+impl std::error::Error for FromEnvError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match &self.inner {
+            #[cfg(any(unix, windows))]
+            FromEnvErrorInner::CannotOpenPath(_, err) => Some(err),
+            #[cfg(unix)]
+            FromEnvErrorInner::NotAPipe(_, Some(err)) | FromEnvErrorInner::CannotOpenFd(_, err) => {
+                Some(err)
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum FromEnvErrorInner {
+    NoEnvVar,
+    CannotParse(String),
+    #[cfg(any(unix, windows))]
+    CannotOpenPath(String, std::io::Error),
```

Comment:
`#[cfg(unix)]` on `CannotOpenFd`/`NotAPipe` allow using `RawFd`, but cfgs on this variant and `Unsupported` do not bring any benefit.